Hide subthreads by blocked users when looking at a post's descendants

ClearlyClaire opened the pull request:
This is about the server not returning known subthreads started by people blocked by the **post's author**, not the viewer.

e.g., if we have `A → B → C → D → E`, where `a → b` means “b replies to a”, and `B` has blocked `D`, querying the descendants for `A` will return `A → B → C`

This affects both the public pages and the API response for the “context” call.

Additionally, do not forward incoming replies when the person being replied to has blocked the person replying, and skip processing incoming posts altogether if the only reason they would have been processed is because they mentioned a local user and that user is blocking the author.

ClearlyClaire committed:
Change threading context code to filter out blocked replies on view

ClearlyClaire committed:
Change ActivityPub processing code to not forward blocked replies

ClearlyClaire committed:
Reword description of some existing tests

ClearlyClaire committed:
Add tests

ClearlyClaire committed:
Spare processing incoming messages that would only be processed because they are mentioning blocked users

trwnh commented:
Would this fix #10916 if I read this correctly?

ClearlyClaire commented:
@trwnh hm, only partially I think. I overlooked the replies collection, let me have a look at it.

ClearlyClaire committed:
Remove blocked users from ActivityPub replies collection

ClearlyClaire committed:
Refactor ActivityPub::RepliesController tests

ClearlyClaire committed:
Add tests

ClearlyClaire commented:
@trwnh pushed changes to remove from replies.

It does not 100% prevent interactions from remote users from being processed, as I am somewhat uncomfortable with that given that:
- there is no real specification for S2S `Block` activities and some implementations ignore them
- not all the local recipients of a remote activity may have blocked the remote account
- I think hiding replies from blocked users should be retroactive, so this behavior should not rely on not processing incoming activities

However, there is still value in skipping processing some activities from blocked users altogether:
- limit wasteful computation and storage
- limit the reach of offensive messages

trwnh commented:
That's fine, I don't expect 100% prevention, I just expect locally blocked users to not be reflected in local responses. However, I do have a response to some points mentioned:

> there is no real specification for S2S Block activities and some implementations ignore them

I don't think S2S block is the correct way to go, I think Blocks should be primarily local, which is to say that it does not matter what other implementations send your server -- your server has the right to filter out or drop certain activities or interactions based on whether the actor is blocked or not.

> not all the local recipients of a remote activity may have blocked the remote account

I don't think other local recipients matter here -- if I am A and I don't want B to reply to me, why should it matter what C, D, or E think? C, D, and E may view the reply on B's profile on B's server, perhaps (if B is remote), but asking A to list replies should not show B's reply. Note that I am taking "subthread" to mean "B directly replied to A, while A blocks B".

> I think hiding replies from blocked users should be retroactive

Nothing in ActivityPub is "retroactive", we're using a delivery-based model with activities. It's like asking to unsend a post after it has already been sent -- you can only send a Delete or Undo. If we're going to take retroactivity into account, then that implies a much more far-reaching change in paradigm, including retroactive followers-only posts for example.

Which is to say: I don't think retroactivity matters here. It's fine for activities to be discarded if they are not allowed/authorized. It's fine to not process them. If you unblock someone, then it's fine to process all future activities without worrying about past activities that were discarded.

>  there is still value in skipping processing some activities from blocked users altogether

I think the value outweighs any potential concerns and it also makes more sense, so it is "more correct". Spec-wise, the original expectation was `The server SHOULD prevent the blocked user from interacting with any object posted by the actor. Servers SHOULD NOT deliver Block Activities to their object. ` -- I am aware that S2S Block is a thing, but I also am in favor of revisiting "stealth blocks" as an option for users (i.e., a toggle to not federate the Block activity, if the user believes the remote server will not respect it or will intentionally disrespect it)

ClearlyClaire commented:
> > not all the local recipients of a remote activity may have blocked the remote account
> 
> I don't think other local recipients matter here -- if I am A and I don't want B to reply to me, why should it matter what C, D, or E think? C, D, and E may view the reply on B's profile on B's server, perhaps (if B is remote), but asking A to list replies should not show B's reply. Note that I am taking "subthread" to mean "B directly replied to A, while A blocks B".

I mean, it kinda matters? If B doesn't know they are blocked by A and want to reply to both B and C, the reply being dropped would be completely unexpected on their part (or even C's if they don't know that A blocks B) and inconsistent.

> > I think hiding replies from blocked users should be retroactive
> 
> Nothing in ActivityPub is "retroactive", we're using a delivery-based model with activities. It's like asking to unsend a post after it has already been sent -- you can only send a Delete or Undo. If we're going to take retroactivity into account, then that implies a much more far-reaching change in paradigm, including retroactive followers-only posts for example.

I'm not talking about dropping activities or anything ActivityPub here, I'm talking about hiding blocked users from replies. If you get harassed by someone replying shit under your replies, you want as much as possible that the replies are not listed, at least from your public pages, even if you blocked them *after* those replies were posted.

> > there is still value in skipping processing some activities from blocked users altogether
> 
> I think the value outweighs any potential concerns and it also makes more sense, so it is "more correct". Spec-wise, the original expectation was `The server SHOULD prevent the blocked user from interacting with any object posted by the actor. Servers SHOULD NOT deliver Block Activities to their object. ` -- I am aware that S2S Block is a thing, but I also am in favor of revisiting "stealth blocks" as an option for users (i.e., a toggle to not federate the Block activity, if the user believes the remote server will not respect it or will intentionally disrespect it)

Again, I'm in favor of skipping processing activities for which *none of the local recipients* block the author, but I'm uneasy with skipping processing activities for which *any of the local recipients* block the author (see reasoning in the first part of this comment). As for stealth blocks, well, I maintained #11562 for nearly two years, but I don't think @Gargron wants that in.

ClearlyClaire commented on app/lib/activitypub/activity/create.rb:
```diff
@@ -402,15 +406,17 @@ def addresses_local_accounts?
     return false if local_usernames.empty?
 
-    Account.local.where(username: local_usernames).exists?
+    scope = Account.local.where(Account.arel_table[:username].lower.in(local_usernames.map(&:downcase)))
+    scope = scope.where.not('EXISTS (SELECT 1 FROM blocks WHERE account_id = accounts.id AND target_account_id = ?)', @account.id)
+    scope.exists?
```
There is a slight (probably pretty much unobservable) overhead to checking for blocks, but the overall query is actually much faster by avoiding a sequential scan (`Account.local.where(username: local_usernames)` did not use the index we have, because the index is on lowered usernames).

ClearlyClaire commented:
Unfortunately, in extreme cases (fabricated case of 4096-statuses-long thread with thousands of participants each blocking thousands of non-participating accounts), the overhead in the `descendant_statuses` CTE makes the query significantly slower (from ~220ms to ~2300ms on my test machine). This is a worst-case scenario though, and this heavily depends on the number of unique participants.

ClearlyClaire committed:
Deduplicate unique accounts and limit account array in descendants CTE

Limit unique accounts to check blocks from to 30 unique accounts (any new
participant will have its blocks ignored for the purpose of displaying the
posts). This is extremely unlikely to be reached in a legitimate situation,
but would limit wasteful computation in specially-crafted threads with many
participants.

ClearlyClaire commented:
ping @Gargron

ClearlyClaire commented:
@Gargron any thoughts on this?

mjankowski commented:
This is two years old and both needs rebase and seems stalled on review.

Is this still viable/desirable? I can take pass at review if so, just want to be sure.

ClearlyClaire commented:
> This is two years old and both needs rebase and seems stalled on review.
> 
> Is this still viable/desirable? I can take pass at review if so, just want to be sure.

I think the behavior change is worthwhile, if incomplete. But this is not for me to chose.

In case we still want it, I rebased it (but have not spent time fixing the linting issues).

graue commented:
@Gargron, please allow this to be merged! It's incredibly frustrating that when I get an abusive reply, there's nothing I can do to avoid providing a platform for it - short of deleting my own post that the person replied to. This is such an easy win that will help prevent people from being forced off of Mastodon due to toxic harassment.

ClearlyClaire committed:
Fix spec linting issues

mjankowski commented:
This has been rebased/maintained for ~2.5 years, which is somewhat absurd.

Is there anything I can do review-wise to help nudge this forward and/or decide to kill it ... or are just waiting on reviewer availability from team?

rkingett commented:
This needs to be merged!

alycecil commented:
Please? I like the default ui but am thinking of switching what my instances run because of creeps reply guying

mjankowski commented:
@mastodon/backend-engineering is this dead, or still planned?

Is code review useful? Is there another blocker?

KeraDamo commented:
> @Gargron, please allow this to be merged! It's incredibly frustrating that when I get an abusive reply, there's nothing I can do to avoid providing a platform for it - short of deleting my own post that the person replied to. This is such an easy win that will help prevent people from being forced off of Mastodon due to toxic harassment.

I still would love to see a fix for this. Harassment is real and I still can 't moderate my own posts, so certain people can just say anything they want. A small improvement would already be to make a post, where comments are turned off as an option. Just like within the ATProtocol.
